models/asymkey/ssh_key_verify.go

```diff
@@ -30,7 +30,9 @@ func VerifySSHKey(ctx context.Context, ownerID int64, fingerprint, token, signat
 		return "", ErrKeyNotExist{}
 	}
 
-	if err := sshsig.Verify(bytes.NewBuffer([]byte(token)), []byte(signature), []byte(key.Content), "gitea"); err != nil {
+	// edge case for Windows based shells that will add CR LF if piped to ssh-keygen command
+	errcrlf := sshsig.Verify(bytes.NewBuffer([]byte(token+"\r\n")), []byte(signature), []byte(key.Content), "gitea")
+	if err := sshsig.Verify(bytes.NewBuffer([]byte(token)), []byte(signature), []byte(key.Content), "gitea"); err != nil && errcrlf != nil {
```

Comment:
Maybe it could be like this:

1. Only try to verify for the original message
2. If it fails to verify the original message, then try to verify "windows style", then clear the error on success.

Then it doesn't need duplicate `sshsig.Verify` for POSIX users.

```suggestion
	err := sshsig.Verify(bytes.NewBuffer([]byte(token)), []byte(signature), []byte(key.Content), "gitea")
	if err != nil {
		// edge case for Windows based shells that will add CR LF if piped to ssh-keygen command
		if sshsig.Verify(bytes.NewBuffer([]byte(token+"\r\n")), []byte(signature), []byte(key.Content), "gitea") == nil {
			err = nil
		}
	}
	if err != nil {
```

Reply:
Thanks for your suggestion @wxiaoguang I applied your commit into the PR. It clearly makes sense to reduce the verification call and attempt it only if the plain (un-CRLFed) token fails.

Another solution that comes to my mind would be the following code:

```go
	if err := sshsig.Verify(bytes.NewBuffer([]byte(token)), []byte(signature), []byte(key.Content), "gitea"); err != nil &&
		sshsig.Verify(bytes.NewBuffer([]byte(token+"\r\n")), []byte(signature), []byte(key.Content), "gitea") != nil {
		log.Error("Unable to validate token signature. Error: %v", err)
		return "", ErrSSHInvalidTokenSignature{
			Fingerprint: key.Fingerprint,
		}
	}
```

This would reduce the additional block for the second if statement.

What do you think?

Reply:
Personally I do not like making code too compact, it's difficult to read or maintain.